python/cac_tripplanner/destinations/views.py

```diff
@@ -391,3 +395,81 @@ def get(self, request, *args, **kwargs):
 
         response = {'destinations': destinations, 'events': events}
         return HttpResponse(json.dumps(response), 'application/json')
+
+
+def return_400(message, error):
+    # Helper to return JSON error messages in a consistent format
+    error = json.dumps({
+        'msg': message,
+        'error': error
+    })
+    return HttpResponse(error, 'application/json', status=400)
+
+
+class UserFlagView(View):
+    """POST-only endpoint for recording anonymous user flags on destinations from mobile app.
+
+    Expects POST as JSON with the following fields:
+     - api_key: String key that must match setting from server secrets file
+     - attraction: ID of the destination or event to flag
+     - is_event: true if attraction is an event
+     - user_uuid: UUID to anonymously identify user (not associated with anything else)
+     - flag: how user has flagged the attraction; must be one of `UserFlag.UserFlags` keys
+             (or empty, if unset)
+    """
+    @method_decorator(csrf_exempt)
+    def dispatch(self, request, *args, **kwargs):
+        return super(UserFlagView, self).dispatch(request, *args, **kwargs)
+
+    def post(self, request, *args, **kwargs):
+        # decode posted body for python 3 support
+        try:
+            json_data = json.loads(request.body.decode('utf-8'))
+        except ValueError as e:
+            return return_400('Could not parse JSON', str(e))
+
+        if json_data.get('api_key') != settings.USER_FLAG_API_KEY:
+            return return_400('api_key required', 'API key missing or incorrect')
+
+        expected_fields = ('attraction', 'is_event', 'user_uuid', 'flag')
+        user_flag_data = {}  # build an object of the cleaned values to use to create flag
+        missing_fields_msg = 'Missing expected value(s): '
+        missing = False  # flag to check if any of the required fields are missing
+        for fld in expected_fields:
+            if fld in json_data:
+                user_flag_data[fld] = json_data.get(fld, None)
+            else:
+                missing = True
+                missing_fields_msg += fld + ', '
+        if missing:
+            return return_400(missing_fields_msg, 'Missing required value')
+
+        # parse string JSON value to boolean
+        user_flag_data['is_event'] = user_flag_data['is_event'] == 'true'
+
+        try:
+            attraction_id = user_flag_data.pop('attraction')
+            if user_flag_data['is_event']:
+                attraction = Event.objects.get(pk=attraction_id)
+            else:
+                attraction = Destination.objects.get(pk=attraction_id)
+
+            if not attraction:
+                return return_400('No attraction found that matches given ID ' + str(attraction_id),
+                                  'Attraction not found')
```

Comment:
Given the way that `.get(...)` works it isn't possible for this block to be reached - it throws either `DoesNotExist` or `MultipleObjectsReturned` depending on the error, and if it succeeds always returns a value.